### ISSUE TITLE
Compression levels for Parquet Compression

### ISSUE THREAD

**Issue by TurnOfACard** [jorgecarleitao/parquet2#119]
Would there be a desire to allow compression levels for parquet compression?

For example, to enable Zstd compression, an option for the interface could be something like:

```rust
pub enum Compression {
    Uncompressed,
    Snappy,
    Gzip,
    Lzo,
    Brotli,
    Lz4,
    Zstd(ZstdLevel),
}

pub struct ZstdLevel(u8);

// Maintains 'compatibility'
impl Default for ZstdLevel {
    fn default() -> Self {
         Self(1)
    }
}

impl ZstdLevel {
    pub fn try_new(level: u8) -> Result<Self, ()>;
}
```

However, that would be a breaking change. I'd be happy to set it up and open a PR if this seems worthwhile.

**Comment by jorgecarleitao** [comment]
I think it is an excellent idea and design.

What do you think about using `Zstd(Option<u8>)` instead, where the `None` represents the default. The advantage is that we can use `Zstd(None)` when reading the file (since the compression ratio is stored in zstd itself and unknown from the file's metadata).

**Comment by TurnOfACard** [comment]
> 

Yes, that makes sense, if `Zstd(Option<ZstdLevel>)` was used then there would be no failure when encoding (i.e. due to an unsupported level like 100 being selected).

It would still allow the end-user to use `Zstd(None)` or `Zstd(Default::default())` if they are ambivalent towards compression levels.

**Comment by kornholi** [comment]
Initially I was on the same track, but perhaps it would be better add a `compression_level: Option<i32>` to `Compressor` and friends instead? or even a `CompressionSettings` enum? It felt awkward tacking it onto `Compression` when it's not used for the metadata or decompression.

**Comment by TurnOfACard** [comment]
> Initially I was on the same track, but perhaps it would be better add a `compression_level: Option<i32>` to `Compressor` and friends instead? or even a `CompressionSettings` enum? It felt awkward tacking it onto `Compression` when it's not used for the metadata or decompression.

That's also a possibility, but if all options were exposed for other compression types (i.e. brotli), would `CompressionSettings` also be an enum? That could result in an error if mismatched `Compressor` and `CompressionSettings` were passed?

**Comment by kornholi** [comment]
Sure, in my example your proposed change would become a separate `CompressionSettings` enum with all the possible settings for each compressor, and we could have `Into<CompressionSettings> for Compression` for the default case. `Compressor` and `compress()` then would take a `CompressionSettings` instead of a `Compression`. It's a nice way not to expose the compression-side details past where they are needed.

**Comment by jorgecarleitao** [comment]
I agree that having separate a enum for reading and writing makes it easier to follow / understand.

**Comment by jorgecarleitao** [comment]
Closed by #128
